File: app/src/main/java/com/example/dandelion/EditProfileActivity.java
package com.example.dandelion;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.ContentResolver;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.StorageTask;
import com.rengwuxian.materialedittext.MaterialEditText;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;

import java.util.Objects;

public class EditProfileActivity extends AppCompatActivity {

    ImageView close, avatar;
    TextView save, tv_change;
    MaterialEditText username, bio;

    FirebaseUser currentUser;

    private Uri mImageUri;
    private StorageTask uploadTask;
    private StorageReference storageReference;
    private FirebaseFirestore mFirestore;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_profile);

        if (getSupportActionBar() != null){
            getSupportActionBar().hide();
        }

        close = findViewById(R.id.close);
        avatar = findViewById(R.id.avatar);
        save = findViewById(R.id.save);
        tv_change = findViewById(R.id.tv_change);
        username = findViewById(R.id.username);
        bio = findViewById(R.id.bio);

        currentUser = FirebaseAuth.getInstance().getCurrentUser();
        storageReference = FirebaseStorage.getInstance().getReference("uploads");
        mFirestore = FirebaseFirestore.getInstance();

        DocumentReference docRef = mFirestore.collection("users").document(currentUser.getUid());
        docRef.addSnapshotListener(new EventListener<DocumentSnapshot>() {
            @Override
            public void onEvent(@Nullable DocumentSnapshot value, @Nullable FirebaseFirestoreException error) {
                if (error != null) {
                    System.err.println("Listen failed: " + error);
                    return;
                }

                if (value != null && value.exists()) {
                    System.out.println("Current data: " + value.getData());
                    User user = value.toObject(User.class);
                    username.setText(user.getUsername());
                    bio.setText(user.getPersona());
                    //Glide.with(getApplicationContext()).load(user.getAvatarImageUrl()).into(mImageUri);

                } else {
                    System.out.print("Current data: null");
                }
            }

        });

        close.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        tv_change.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                /*CropImage.activity()  todo...
                        .setAspectRatio(1, 1)
                        .setCropShape(CropImageView.CropShape.OVAL)
                        .start(EditProfileActivity.this);*/
            }
        });

        save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateProfile(username.getText().toString(), bio.getText().toString());
                finish();
            }
        });

        avatar.setOnClickListener(new View.OnClickListener() {
            @RequiresApi(api = Build.VERSION_CODES.KITKAT)
            @Override
            public void onClick(View v) {
                updateProfile(Objects.requireNonNull(username.getText()).toString(),
                        Objects.requireNonNull(bio.getText()).toString());
                
            }
        });
    }

    private void updateProfile(final String username, final String bio) {
        DocumentReference docRef = mFirestore.collection("users").document(currentUser.getUid());
        docRef.get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
            @Override
            public void onSuccess(DocumentSnapshot documentSnapshot) {
                //TODO: so far only accept one persona
                mFirestore.collection("users").document(currentUser.getUid())
                        .update(
                                "username", username,
                                "persona", bio
                        );

            }
        });
    }

    private String getFileExtention(Uri uri){
        ContentResolver contentResolver = getContentResolver();
        MimeTypeMap mimeTypeMap = MimeTypeMap.getSingleton();
        return mimeTypeMap.getExtensionFromMimeType(contentResolver.getType(uri));
    }

    private void uploadImage(){
        if (mImageUri != null){
            final StorageReference filereference = storageReference.child(System.currentTimeMillis()
            +"."+getFileExtention(mImageUri));

            uploadTask = filereference.putFile(mImageUri);
            uploadTask.continueWithTask(new Continuation() {
                @Override
                public Object then(@NonNull Task task) throws Exception {
                    if (!task.isSuccessful()){
                        throw task.getException();
                    }
                return filereference.getDownloadUrl();
                }
            }).addOnCompleteListener(new OnCompleteListener<Uri>() {
                @Override
                public void onComplete(@NonNull Task<Uri> task) {
                    if (task.isSuccessful()){
                        Uri downloadUri = task.getResult();
                        final String myUrl = downloadUri.toString();

                        DocumentReference docRef = mFirestore.collection("users").document(currentUser.getUid());
                        docRef.get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                            @Override
                            public void onSuccess(DocumentSnapshot documentSnapshot) {
                                //TODO: so far only accept one persona
                                mFirestore.collection("users").document(currentUser.getUid())
                                        .update(
                                                "avatarImageUri", myUrl
                                        );

                            }
                        });


                    }
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    Toast.makeText(EditProfileActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        } else{
            Toast.makeText(this, "No image Selected", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE && resultCode == RESULT_OK){
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            mImageUri = result.getUri();

            uploadImage();
        }else{
            Toast.makeText(this, "Something gone wrong", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/unlone/ui/profile/EditProfileActivity.kt
package com.example.unlone.ui.profile

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.example.unlone.R
import com.example.unlone.databinding.ActivityEditProfileBinding
import com.example.unlone.instance.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.toObject
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.rengwuxian.materialedittext.MaterialEditText
import java.util.*

class EditProfileActivity : AppCompatActivity() {
    var username: MaterialEditText? = null
    var bio: MaterialEditText? = null
    var currentUser: FirebaseUser? = null
    private var storageReference: StorageReference? = null
    private var mFirestore: FirebaseFirestore? = null

    private lateinit var binding: ActivityEditProfileBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_profile)
        if (supportActionBar != null) {
            supportActionBar!!.hide()
        }

        binding = ActivityEditProfileBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        val cancel = binding.cancelButton
        val save = binding.saveButton
        val username = binding.username
        val bio = binding.bio
        val likes = binding.likes
        val dislikes = binding.dislikes

        cancel.setOnClickListener{ finish() }
        save.setOnClickListener{
            updateProfile(username.text.toString(), bio.text.toString())
            finish()
        }

        currentUser = FirebaseAuth.getInstance().currentUser
        storageReference = FirebaseStorage.getInstance().getReference("uploads")
        mFirestore = FirebaseFirestore.getInstance()
        val docRef = mFirestore!!.collection("users").document(
            currentUser!!.uid
        )
        docRef.addSnapshotListener(EventListener { value, error ->
            if (error != null) {
                System.err.println("Listen failed: $error")
                return@EventListener
            }
            if (value != null && value.exists()) {
                println("Current data: " + value.data)
                val user = value.toObject<User>()
                username.setText(user!!.username)
                bio.setText(user.bio)
                // TODO ("set bio, likes, and dislikes into persona")

            } else {
                print("Current data: null")
            }
        })
    }

    private fun updateProfile(username: String, bio: String) {
        val docRef = mFirestore!!.collection("users").document(
            currentUser!!.uid
        )
        docRef.get().addOnSuccessListener { //TODO: so far only accept one persona
            mFirestore!!.collection("users").document(currentUser!!.uid)
                .update(
                    "username", username,
                    "bio", bio
                // TODO (set likes and dislikes)
                )
        }
    }
}